### PR TITLE
RATIFY and GOV constraint tests

### LINKED ISSUE #3784: Constraints and properties for GOV rule
Constraint based testing framework is mature enough to support Conway.

We need to add properties for the GOV rule

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
# Description

A bunch more work on constraint-based STS testing. This time for running the RATIFY and GOV rules. The GOV rule doesn't work yet because we need to extend the constraint language a bit to handle the map-of-maps in the GOV signal.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/input-output-hk/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/input-output-hk/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs
Comment (lehins) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/STS.hs]: You probably wanna see where compaction fails, if it does fail.
```suggestion
toCompF :: (HasCallStack, Compactible v) => Map.Map k v -> Map.Map k (UM.CompactForm v)
```

**Review by lehins — APPROVED** [review]
Looks great. Thank you!

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs]: You probably want to break this up, because in pre conway eras some VState things are empty

```
vstatePreds p = case whichPParams p of
  PParamsConwayToConway ->
    [ Sized (Range 10 20) currentDRepState
    , Sized (Range 5 7) (Dom committeeState)
.. the deposit field of the currentDRepState is the drepDeposit
   ]
  _ ->
    [ Sized (Range 0 0) currentDRepState
    , Sized (Range 0 0) committeeState
    ]
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs]: Just curious, where do the  Show (TxF era), Show (TxBodyF era)  arise?

**Review by TimSheard — CHANGES_REQUESTED** [review]
Why do the PrettyA constraints pop up everywhere? I will help figure out how### to make them go away.

**Comment by MaximilianAlgehed** [comment]
@TimSheard most of the `PrettyA` constraints come from this line:
```haskell
instance (EraPParams era, PrettyA (PParamsUpdate era)) => PrettyA (ProposalProcedure era) where
```

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs
Comment (MaximilianAlgehed) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs]: Which of these are generation constraints and which are checking constraints?

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs
Comment (MaximilianAlgehed) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs]: They come from these two lines:
```haskell
synopsis (TxBodyR _) x = show x
...
synopsis (TxR _) x = show x
```

Comment (MaximilianAlgehed) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs]: I wish we could get rid of them but the problem is that they (and `PParamsUpdates`) are era-indexed

Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs]: Do this


Change these
synopsis (TxBodyR _) x = show x
synopsis (TxR _) x = show x

To
synopsis (TxBodyR p) x = show (pcTxBody p x)
synopsis (TxR p) x = show (pcTx p x)

add these imports
import Test.Cardano.Ledger.Generic.PrettyCore (pcTxBody,pcTx)

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs
Comment (TimSheard) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs]: I guess
these two
 [ Sized (Range 0 0) currentDRepState
    , Sized (Range 0 0) committeeState
    ]
are checking, since they need to be empty before Conway Era

There is some checking constrain, implied by this comment
... he deposit field of the currentDRepState is the drepDeposit
But I am not sure how to best express this. It goes something like this. Add up all the deposit fields of the 'currentDRepState' this must be less then the utxosDeposit field in the utxoState. It is more prcise than that, since there are 4 places deposits are stored, and the sum of all 4 must be equal to the variable 'deposits'

**Comment by TimSheard** [comment]
We can fix this
instance (EraPParams era, PrettyA (PParamsUpdate era)) => PrettyA (ProposalProcedure era) where
by adding a name function for pretty printing PParams as is done here in this PR
https://github.com/input-output-hk/cardano-ledger/tree/ts-constrained-trace
I wish it didn't take so long to get PRs merged

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs
Comment (MaximilianAlgehed) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs]: I get the following from the compiler if I do this:
```
src/Test/Cardano/Ledger/Constrained/TypeRep.hs:588:43: error:
    ΓÇó Could not deduce: Core.TxBody e ~ TxBodyF e
      from the context: (t ~ TxBodyF e, Era e)
        bound by a pattern with constructor:
                   TxBodyR :: forall era.
                              Era era =>
                              Proof era -> Rep era (TxBodyF era),
                 in an equation for ΓÇÿsynopsisΓÇÖ
        at src/Test/Cardano/Ledger/Constrained/TypeRep.hs:588:11-19
      Expected: Core.TxBody e
        Actual: t
    ΓÇó In the second argument of ΓÇÿpcTxBodyΓÇÖ, namely ΓÇÿxΓÇÖ
      In the first argument of ΓÇÿshowΓÇÖ, namely ΓÇÿ(pcTxBody p x)ΓÇÖ
      In the expression: show (pcTxBody p x)
    ΓÇó Relevant bindings include
        p :: Proof e
          (bound at src/Test/Cardano/Ledger/Constrained/TypeRep.hs:588:19)
        synopsis :: Rep e t -> t -> String
          (bound at src/Test/Cardano/Ledger/Constrained/TypeRep.hs:508:1)
    |
588 | synopsis (TxBodyR p) x = show (pcTxBody p x)
    |                                           ^

src/Test/Cardano/Ledger/Constrained/TypeRep.hs:593:35: error:
    ΓÇó Could not deduce: Core.Tx e ~ TxF e
      from the context: (t ~ TxF e, Era e)
        bound by a pattern with constructor:
                   TxR :: forall era. Era era => Proof era -> Rep era (TxF era),
                 in an equation for ΓÇÿsynopsisΓÇÖ
        at src/Test/Cardano/Ledger/Constrained/TypeRep.hs:593:11-15
      Expected: Core.Tx e
        Actual: t
    ΓÇó In the second argument of ΓÇÿpcTxΓÇÖ, namely ΓÇÿxΓÇÖ
      In the first argument of ΓÇÿshowΓÇÖ, namely ΓÇÿ(pcTx p x)ΓÇÖ
      In the expression: show (pcTx p x)
    ΓÇó Relevant bindings include
        p :: Proof e
          (bound at src/Test/Cardano/Ledger/Constrained/TypeRep.hs:593:15)
        synopsis :: Rep e t -> t -> String
          (bound at src/Test/Cardano/Ledger/Constrained/TypeRep.hs:508:1)
    |
593 | synopsis (TxR p) x = show (pcTx p x)
    |                                   ^
```

Comment (MaximilianAlgehed) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs]: If I change it to `synopsis (TxBodyR p) x = show (pcTxBody p (unTxBodyF x))` and `synopsis (TxR p) x = show (pcTx p (unTxF x))` I instead get:
```
src/Test/Cardano/Ledger/Constrained/TypeRep.hs:589:32: error:
    ΓÇó Could not deduce (Cardano.Ledger.Pretty.PrettyA
                          (Core.PParamsUpdate e))
        arising from a use of ΓÇÿpcTxBodyΓÇÖ
      from the context: (t ~ TxBodyF e, Era e)
        bound by a pattern with constructor:
                   TxBodyR :: forall era.
                              Era era =>
                              Proof era -> Rep era (TxBodyF era),
                 in an equation for ΓÇÿsynopsisΓÇÖ
        at src/Test/Cardano/Ledger/Constrained/TypeRep.hs:589:11-19
    ΓÇó In the first argument of ΓÇÿshowΓÇÖ, namely
        ΓÇÿ(pcTxBody p (unTxBodyF x))ΓÇÖ
      In the expression: show (pcTxBody p (unTxBodyF x))
      In an equation for ΓÇÿsynopsisΓÇÖ:
          synopsis (TxBodyR p) x = show (pcTxBody p (unTxBodyF x))
    |
589 | synopsis (TxBodyR p) x = show (pcTxBody p (unTxBodyF x))
    |                                ^^^^^^^^

src/Test/Cardano/Ledger/Constrained/TypeRep.hs:594:28: error:
    ΓÇó Could not deduce (Cardano.Ledger.Pretty.PrettyA
                          (Core.PParamsUpdate e))
        arising from a use of ΓÇÿpcTxΓÇÖ
      from the context: (t ~ TxF e, Era e)
        bound by a pattern with constructor:
                   TxR :: forall era. Era era => Proof era -> Rep era (TxF era),
                 in an equation for ΓÇÿsynopsisΓÇÖ
        at src/Test/Cardano/Ledger/Constrained/TypeRep.hs:594:11-15
    ΓÇó In the first argument of ΓÇÿshowΓÇÖ, namely ΓÇÿ(pcTx p (unTxF x))ΓÇÖ
      In the expression: show (pcTx p (unTxF x))
      In an equation for ΓÇÿsynopsisΓÇÖ:
          synopsis (TxR p) x = show (pcTx p (unTxF x))
    |
594 | synopsis (TxR p) x = show (pcTx p (unTxF x))
    |                            ^^^^
Failed, 34 modules loaded.
```

Comment (MaximilianAlgehed) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/Tx.hs]: Which is less than helpful, we are back in the PParamsUpdate quagmire

**Comment by MaximilianAlgehed** [comment]
> We can fix this instance (EraPParams era, PrettyA (PParamsUpdate era)) => PrettyA (ProposalProcedure era) where by adding a name function for pretty printing PParams as is done here in this PR https://github.com/input-output-hk/cardano-ledger/tree/ts-constrained-trace I wish it didn't take so long to get PRs merged

Can you please provide a pointer to the relevant code?

**Comment by MaximilianAlgehed** [comment]
@TimSheard another solution to this is to add a `type TestableEra era = (Reflect era, PrettyA (Core.PParamsUpdate era))` constraint that we have everywhere. I think this would be fairly non-intrusive (all eras that you actually want to use will satisfy this constraint) and it would let us add things here that we need without having tonnes of ugly constraints floating around the codebase - instead all our testing functions would require `TestableEra era` and that would be it.

**Comment by lehins** [comment]
> type TestableEra era = (Reflect era, PrettyA (Core.PParamsUpdate era))

I'd advise against using constraint type synonyms. Almost every time we used them it lead to problems, since compiler cannot advise you on redundant constraints, leading to over constraining in one place or another.

That being said this is a self contained package that no-one depends on and it is a pretty benign synonym, so I am not gonna block you from using it.

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs
Comment (MaximilianAlgehed) [reply libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/CertState.hs]: I've done this but I think we'll leave the details of implementing the constraints for the next PR - maybe after we've gotten some more input from the team on the constraints.

**Comment by MaximilianAlgehed** [comment]
@tim

**Comment by MaximilianAlgehed** [comment]
@TimSheard do you have a pointer to the PParamsUpdate fix?

**Comment by MaximilianAlgehed** [comment]
@TimSheard do you mind rebasing this? I'm not up to speed with all the edits so it would be good to have your eye on the rebase so I don't screw something up.

**Comment by lehins** [comment]
@MaximilianAlgehed We prefer rebase instead of commits that merge master into ledger

This results in a nice linear history.

In particular I am talking about this commit: https://github.com/input-output-hk/cardano-ledger/pull/3792/commits/344ca7d157d80647af6a54138cdfeb41a9c68b04

**Comment by MaximilianAlgehed** [comment]
@lehins this will be squashed before merging anyway.

**Comment by MaximilianAlgehed** [comment]
The problem is that rebasing when the history is a mess of rebases causes more problems than solutions.

**Comment by lehins** [comment]
> The problem is that rebasing when the history is a mess of rebases causes more problems than solutions.


I prefer to squash and then rebase. It is easier to resolve conflicts that way. Whatever works for you though. I just wanted to make sure we don't get those kind of merge commits into master, that's all.

**Comment by MaximilianAlgehed** [comment]
> > The problem is that rebasing when the history is a mess of rebases causes more problems than solutions.
> 
> I prefer to squash and then rebase. It is easier to resolve conflicts that way. Whatever works for you though. I just wanted to make sure we don't get those kind of merge commits into master, that's all.

Yes, that's what I usually do too. But here this was rebased on top of another PR before that PR was merged - which causes the history to be a bit messy...

**Comment by MaximilianAlgehed** [comment]
@lehins I think there is a CI bug in 8.10.7 causing the build failure - can you restart it?

**Comment by lehins** [comment]
> I think there is a CI bug in 8.10.7 causing the build failure - can you restart it?

Most certainly. We gotta wait for CI to finish before GithubActions provides the ability to restart failed builds.

>  But here this was rebased on top of another PR before that PR was merged - which causes the history to be a bit messy

You might know about this, but I found this pretty cool, so I'll share it with you. If I understand the problem you mentioned correctly then there is a simple solution in general.

Let's say you have this history:
```
master:
A--B--C

branch-a:
A--D

branch-b (eg. branch in this PR):
A--D--E--F
````

PR for `branch-a` gets rebased and merged into master:
```
master (with branch-a merged in):
A--B--C--D'

branch-b (eg. branch in this PR):
A--D--E--F
```

Then what you do on `branch-b` is `git rebase -i origin/master` and simply drop the `D`  commit and you'll get the expected history with no conflicts that would normally be related to rebasing `D` on commit `D'`:

```
branch-b:
A--B--C--D'--E'--F'
```

**Comment by MaximilianAlgehed** [comment]
The problem is when the other branch has changed _before_ being merged but that change hasn't been merged into your branch...

**Comment by lehins** [comment]
> I think there is a CI bug in 8.10.7 causing the build failure - can you restart it?

@MaximilianAlgehed Looks like tests are not passing. So, there is no point in restarting

**Comment by MaximilianAlgehed** [comment]
@TimSheard any clue what's going on with that test? I don't think this PR should change anything there, but that's probably some rebase artifact.

**Comment by TimSheard** [comment]
Some how the rebase lost a constraint suming the Drep deposits. Fixed it up

File: libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs
Comment (TimSheard) [inline libs/cardano-ledger-test/src/Test/Cardano/Ledger/Constrained/Preds/LedgerState.hs]: I fixed this commented out code.

**Review by TimSheard — COMMENTED** [review]
I found one thing and I fixed it.  Ready to go!

**Review by TimSheard — APPROVED** [review]

**Pull request merged**
